lock-api/src/main/java/com/palantir/lock/client/LeaderElectionReportingTimelockService.java

```diff
@@ -122,6 +140,129 @@ public ConjureStartTransactionsResponse startTransactions(ConjureStartTransactio
         return election && (currentLeader != null);
     }
 
+    /**
+     * Estimating the duration of leader election:
+     *
+     * If a request is sent out at time T_1, and a response arrives at time T_2 with the id of leader A, we can
+     * deduce three things:
+     *   1. A could not have lost leadership before T_1 if it was the leader before it (upper bound)
+     *   2. A must have gained leadership before T_2 (lower bound)
+     *   3. A was the leader at least in some point in the interval [T_1, T_2]
+     * Although from just one request, the lower bound is greater than the upper bound, after multiple responses that
+     * is generally not going to be the case anymore.
```

Comment:
IMO this is pretty hard to take in. My understanding of it is as follows:
0. Lower and upper bound are referring to the temporal bounds of when A is the leader (I don't think that's really mentioned explicitly, and it's not obvious just from the text here).
1. Lower bound is the first successful response from leader A;
2. Upper bound is the last successful request to leader A.

While what you've said is true for a single request/response pair, it's quite unintuitive reasoning on that scale.

Reply:
It's literally said in 1 and 2:
A could not have lost leadership before T_1
A must have gained leadership before T_2

But yeah, maybe upper bound is confusing as it is not an upper bound on A's leadership, it's the best underapproximation of the leadership end time we can make with known information. Same for the lower bound, it's the best overaproximation of the leadership start time that we can make with known information. That is however, quite wordy...

Reply:
Yes, I can read that, but reasoning over a single request/response pair doesn't really make a lot of sense mentally (at least, not to me), even if what you have written is correct. The other thing is that you didn't really say what the bound if of - it could refer to bounding when the time when A was the leader, or refer to the moment when A lost leadership. It might sound obvious, but in my opinion it is best to be as explicit as possible here so that someone without all context can read this with minimal frustration.